File: dash_docs/chapters/dash_canvas/examples/canvas_annotations.R
library(dash)
library(dashHtmlComponents)
library(dashCanvas)
library(dashTable)
library(jsonlite)
library(magrittr)

filename <- 'https://upload.wikimedia.org/wikipedia/commons/e/e4/Mitochondria%2C_mammalian_lung_-_TEM_%282%29.jpg'
canvas_width <- 500
columns <- c("type", "width", "height", "scaleX", "strokeWidth", "path")

app <- Dash$new()

app$layout(htmlDiv(list(
  htmlH6('Draw on image and press Save to show annotations geometry'),
  htmlDiv(list(
    dashCanvas(
      id = 'annot-canvas',
      lineWidth = 5,
      filename = filename,
      width = canvas_width
    )
  )),
  htmlDiv(
    dashDataTable(
      id = 'annot-canvas-table',
      style_table = list('overflowX' = 'scroll'),
      style_cell = list(textAlign = 'left'),
      columns = lapply(columns, function(col) {
        return(list(name = col, id = col))
      })
    )
  )
)))

app$callback(
  output = list(id='annot-canvas-table', property='data'),
  params = list(input(id='annot-canvas',property='json_data')),

  function(json) {
    if (is.null(json[[1]]) || is.na(json[[1]])) {
      return(list(0, 0, 0, 0, 0, 0))
    } else {
      data <- fromJSON(json)
      df <- data[['objects']]
      df_filtered <- df[2:length(df)]
      return(df_filtered)
    }
  }
)

app$run_server()

File: dash_docs/chapters/dash_canvas/examples/canvas_annotations.py
import dash
from dash.exceptions import PreventUpdate
from dash.dependencies import Input, Output, State
import dash_html_components as html
from dash_canvas import DashCanvas
import json
from dash_table import DataTable

app = dash.Dash(__name__)

filename = 'https://raw.githubusercontent.com/plotly/datasets/master/mitochondria.jpg'
canvas_width = 500

columns = ['type', 'width', 'height', 'scaleX', 'strokeWidth', 'path']

app.layout = html.Div([
    html.H6('Draw on image and press Save to show annotations geometry'),
    DashCanvas(id='annot-canvas',
               lineWidth=5,
               filename=filename,
               width=canvas_width,
               ),
    DataTable(id='annot-canvas-table',
              style_cell={'textAlign': 'left'},
              columns=[{"name": i, "id": i} for i in columns]),
    ])


@app.callback(Output('annot-canvas-table', 'data'),
              [Input('annot-canvas', 'json_data')])
def update_data(string):
    if string:
        data = json.loads(string)
    else:
        raise PreventUpdate
    return data['objects'][1:]


if __name__ == '__main__':
    app.run_server(debug=True)
